In [1]:
import sys
print sys.version

2.7.15 |Anaconda, Inc.| (default, May  1 2018, 23:32:55) 
[GCC 7.2.0]


# 1. svm toy

In [2]:
import time
import numpy
from sklearn.datasets import make_blobs, make_gaussian_quantiles, load_digits
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

import warnings; warnings.filterwarnings('ignore')

In [3]:
2. ** numpy.arange(3,-15,-3)

array([8.00000000e+00, 1.00000000e+00, 1.25000000e-01, 1.56250000e-02,
       1.95312500e-03, 2.44140625e-04])

In [4]:
def GridSearch(X_train, y_train, k):
    
    # define range dos parametros
    C_range = 2. ** numpy.arange(-5,15,3)
    gamma_range = 2. ** numpy.arange(3,-15,-3)
    k = numpy.array([k])
    param_grid = dict(gamma=gamma_range, C=C_range, kernel=k)
    
    # instancia o classificador, gerando probabilidades
    srv = SVC(probability=True)
    
    # faz a busca
    grid = GridSearchCV(srv, param_grid, n_jobs=-1, verbose=False)
    grid.fit(X_train, y_train)
    
    # recupera o melhor modelo
    model = grid.best_estimator_
    
    # imprime os parametros desse modelo
    print '\nMelhores parâmetros:', grid.best_params_, '\n'
    return model

## a) make_blobs

In [5]:
X, y = make_blobs(n_samples=300, centers=2)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
for k in ['linear', 'rbf']:
    
    print '\nInicializando o com o kernel %s' % k
    clf = SVC(kernel=k)
    clf.fit(X_train, y_train)
    # mostra o resultado do classificador na base de teste
    print 'Desempenho com Kernel %s:' % k, clf.score(X_test, y_test)
    print 'Vetores de suporte:',  clf.n_support_[0] + clf.n_support_[1]

    # GridSearch retorna o melhor modelo encontrado na busca
    best = GridSearch(X_train, y_train, k=k)

    # resultado do treinamento
    print 'Acurácia com kernel %s:' % k, best.score(X_train, y_train)
    print 'Vetores de suporte:',  best.n_support_[0]+best.n_support_[1]


Inicializando o com o kernel linear
Desempenho com Kernel linear: 1.0
Vetores de suporte: 2

Melhores parâmetros: {'kernel': 'linear', 'C': 0.03125, 'gamma': 8.0} 

Acurácia com kernel linear: 1.0
Vetores de suporte: 4

Inicializando o com o kernel rbf
Desempenho com Kernel rbf: 1.0
Vetores de suporte: 39

Melhores parâmetros: {'kernel': 'rbf', 'C': 0.03125, 'gamma': 0.125} 

Acurácia com kernel rbf: 1.0
Vetores de suporte: 110


O kernel RBF produz mais vetores de suporte que o kernel linear. Ambos SVMs atingem 100% de acurácia na base de teste

## b) make_gaussian_quantiles

In [7]:
X, y = make_gaussian_quantiles(n_samples=300)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
for k in ['linear', 'rbf']:
    
    print '\nInicializando o com o kernel %s' % k
    clf = SVC(kernel=k)
    clf.fit(X_train, y_train)
    # mostra o resultado do classificador na base de teste
    print 'Desempenho com Kernel %s:' % k, clf.score(X_test, y_test)
    print 'Vetores de suporte:',  clf.n_support_[0] + clf.n_support_[1]


Inicializando o com o kernel linear
Desempenho com Kernel linear: 0.45
Vetores de suporte: 159

Inicializando o com o kernel rbf
Desempenho com Kernel rbf: 0.9333333333333333
Vetores de suporte: 97


Com a função make_gaussian_quantiles o kernel linear produz mais vetores de suporte. O kernel RBF produz uma acurácia muito melhor que o linear. O kernel linear produz 156 vetores de suporte enquanto o kernel RBF produz 105 vetores de suporte.

# 2. svm

In [9]:
X = load_digits().data
y = load_digits().target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [10]:
for k in ['linear', 'rbf']:
    
    start = time.time()
    
    print '\nInicializando o com o kernel %s' % k
    clf = SVC(kernel=k)
    clf.fit(X_train, y_train)
    # mostra o resultado do classificador na base de teste
    print 'Desempenho com Kernel %s:' % k, clf.score(X_test, y_test)
    print 'Vetores de suporte:',  clf.n_support_[0] + clf.n_support_[1]

    end = time.time()
    print 'Tempo de treinamento: %s segundos com kernel' % round(end-start, 2), k


Inicializando o com o kernel linear
Desempenho com Kernel linear: 0.9777777777777777
Vetores de suporte: 76
Tempo de treinamento: 0.09 segundos com kernel linear

Inicializando o com o kernel rbf
Desempenho com Kernel rbf: 0.4666666666666667
Vetores de suporte: 299
Tempo de treinamento: 0.91 segundos com kernel rbf


## a)

O melhor kernel em termos de acurácia foi linear com 97% de acerto na base de teste enquanto o kernel RBF acertou 47% na base de teste. O kernel linear foi uma ordem de magnitude mais rápido que o kernel RBF

## b)

O SVM com kernel linear produziu 76 vetores de suporte, enquanto o kernel RBF produziu 299 vetores de suporte